src/IO/WriteBufferFromS3.cpp

```diff
@@ -101,13 +94,18 @@ void WriteBufferFromS3::finalizeImpl()
         writePart();
         completeMultipartUpload();
     }
-
-    finalized = true;
 }
 
 WriteBufferFromS3::~WriteBufferFromS3()
 {
-    finalizeImpl();
+    try
+    {
+        next();
```

Comment:
But why it does not call finalize() anymore? AFAIU it should call finalize() not next()?

Reply:
@azat As I understand, `finalize()` is called before query returns to client, but this call was forgotten when we made compression argument, and behavior of table functions with and without compression became inconsistent:

```
cristal :) insert into table function url('https://s3.us-west-2.amazonaws.com/altinity-qa-testa/data/', 'CSVWithNames', 'd UInt64', 'deflate') SELECT * from numbers(1)

INSERT INTO FUNCTION url('https://s3.us-west-2.amazonaws.com/altinity-qa-testa/data/', 'CSVWithNames', 'd UInt64', 'deflate') SELECT *
FROM numbers(1)

Query id: ca24b2f0-b79e-4a60-91f0-85a7caf9bf66

Ok.

0 rows in set. Elapsed: 0.736 sec. 

cristal :) insert into table function url('https://s3.us-west-2.amazonaws.com/altinity-qa-testa/data/', 'CSVWithNames', 'd UInt64') SELECT * from numbers(1)

INSERT INTO FUNCTION url('https://s3.us-west-2.amazonaws.com/altinity-qa-testa/data/', 'CSVWithNames', 'd UInt64') SELECT *
FROM numbers(1)

Query id: 71a4779b-8da4-4e18-ab21-dd9abac1b1f0

→ Progress: 1.00 rows, 8.00 B (1.20 rows/s., 9.63 B/s.)  99%
0 rows in set. Elapsed: 0.939 sec. 

Received exception from server (version 21.2.1):
Code: 86. DB::Exception: Received from localhost:9000. DB::Exception: Received error from remote server /altinity-qa-testa/data/. HTTP status code: 405 Method Not Allowed, body: <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>MethodNotAllowed</Code><Message>The specified method is not allowed against this resource.</Message><Method>POST</Method><ResourceType>OBJECT</ResourceType><RequestId>M6Q1NHCWGG0F0SD0</RequestId><HostId>EGOFX9bG3aMMVswqXtCTYPJoIQ/jFHOo2AG9d0JgjdvLF+RNYK9TiQCeLgaAAg8OyKFAp8Z0iqg=</HostId></Error>. 

```

Previously, I fixed that with these lines of code, calling `finalize()` in destructor. But it is really a bad place to do so, because it involves lots of initialization code and exceptions in destructor, which shall be catch. `finalize()` of these write buffers is called at `writeSuffix()` of corresponding output stream.

Reply:
>As I understand, finalize() is called before query returns to client

It should, can you give a stacktrace example?

>but this call was forgotten when we made compression argument, and behavior of table functions with and without compression became inconsistent

Does #21305 fixes it?

Reply:
Yes it does.

Reply:
I'll copy tests to another PR.